### PR TITLE
Allow dashboard to run without resource service

### PR DESCRIPTION
**Pull request by drewnoakes**
(Replaces #1795 which closed when we merged the feature branch.)

Relates to #1556

With the dashboard running OOP, it is configured with the resource server endpoint via the `DOTNET_DASHBOARD_GRPC_ENDPOINT_URL` environment variable.

The dashboard can actually perform useful work even without this connection, however it requires some code changes to support this.

In this PR, if the environment variable is not defined, then `IDashboardClient.IsEnabled` will be false, and some code paths become inactive based off the flag.

To get tests passing, we introduce an abstraction for environment variables so we can inject the values we need.

The navigation tree will not display the "Resources" or "Console logs" items when no resource service endpoint is available:

image

---

Pending work:

- [ ] On load the page currently routes to `/` which shows an empty _Resources_ list. Route somewhere else.
- [x] Update dashboard's README

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/1815)

### CODE REVIEW

**Comment by JamesNK** [comment]
The title should have a value when there is no resource server. "Aspire dashboard"?

An option to set it via argument/env var would be useful, but no one has asked for that yet.

File: src/Aspire.Hosting/Utils/IEnvironmentVariables.cs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    /// </summary>
+    private ImmutableDictionary<string, string?> _valueByName = ImmutableDictionary<string, string?>.Empty.WithComparers(StringComparers.EnvironmentVariableName);
+
+    [return: NotNullIfNotNull(nameof(defaultValue))]
+    public string? GetString(string variableName, string? defaultValue = null)
+    {
+        // Environment.GetEnvironmentVariable queries the variable each time,
+        // but our variables don't change during the lifetime of the process.
+        // So we cache them for faster repeat lookup.
+        var value = ImmutableInterlocked.GetOrAdd(ref _valueByName, key: variableName, valueFactory: Environment.GetEnvironmentVariable);
```
[the remark this answers is not in the document]
Comment (davidfowl) [reply src/Aspire.Hosting/Utils/IEnvironmentVariables.cs:37]: Agree we should switch to IConfiguration.

**Comment by drewnoakes** [comment]
> The title should have a value when there is no resource server. "Aspire dashboard"?
> 
> An option to set it via argument/env var would be useful, but no one has asked for that yet.

I like this idea! Added:

```c#
    string IDashboardClient.ApplicationName
    {
        get => _applicationName
            ?? _environmentVariables.GetString("DOTNET_DASHBOARD_APPLICATION_NAME")
            ?? "Aspire";
    }
```

File: src/Aspire.Hosting/Utils/IEnvironmentVariables.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (drewnoakes) [reply src/Aspire.Hosting/Utils/IEnvironmentVariables.cs:37]: Agree. Filed https://github.com/dotnet/aspire/issues/1836 and will fix it in a subsequent PR just to get this merged and under test.

**Comment by drewnoakes** [comment]
@JamesNK PTAL. I'd like to get this merged in so there's more chance it's included in the next round of testing.

**Review by JamesNK — APPROVED** [review]

**Pull request merged**